### PR TITLE
bdBank Credit

### PR DESCRIPTION
**Pull request by datbth**
# Introduction
We want to have a credit system so that some specific changes to users' balances do not affect their user promotions. Example changes:
* Bonus adjustments

# Database changes
* New column in table `xf_user`:
  * bdbank_credit (money): amount of credit of user

# Implementation
New transaction types:
* TYPE_CREDITABLE (3): when a transaction of this new type is made, each involved user is granted an amount of credit points equal to the amount of money that he/she **lost**
* TYPE_ADJUSTMENT (4): adjustment transactions are made when rebuilding bonuses to adjust the total amount of effective points granted to bonusing actions
### Before
1. For each type of bonus
    1. For each bonusing action record
        1. Reverse old transactions
        2. Create new transaction with new bonus amount

### After
1. For each type of bonus
    1. For each bonusing action record
        1. If SUM(total_transaction_amount + total_archived_amount + total_adjusted_amount) != bonus
            1. Insert a new adjustment transaction with amount = bonus - above sum

## User Info Rebuild
After rebuilding all bonuses, rebuild each user's credit by adding:
* the **negated** total amount of their adjustment transactions
* the total amount of the creditable transactions

## User Promotion
Calculate user promotions with `bdbank_money + bdbank_credit`

# Other changes
This PR also enables using environment config to specify bonuses by time period.  
Example:
```php
$config['bdBank_bonusesByTimePeriod'] = array(
    array(
        'start_date' => 1531233353,
        'end_date' => 1531233357,
        'bonuses' => array(
            'liked' => 4
        )
    ),
    array(
        'start_date' => 1531233357,
        'end_date' => 1531233367,
        'bonuses' => array(
            'liked' => 6
        )
    )
);
```
-> Likes received from 1531233353 (inclusive) to 1531233357 (exclusive) will give 4 points, likes received from 1531233357 (inclusive) to 1531233367 (exclusive) will give 6 points, regardless of 'bonus_liked' Option

### CODE REVIEW

File: library/bdBank/Deferred/User.php [hunk 1]
```diff
@@ -69,4 +84,35 @@ protected function _fetchSum($formula, $where)
...
+
+        $totalAdjustedAmount = $db->fetchOne("
+            SELECT SUM(adj.amount)
+            FROM xf_bdbank_transaction_adjustment adj
+                LEFT JOIN xf_bdbank_transaction t
+                    ON adj.comment = t.comment
+                        AND t.reversed = 0
+                LEFT JOIN xf_bdbank_archive a
+                    ON adj.comment = a.comment 
+            WHERE t.$userIdCol = $userId
```
Comment (daohoangson) [inline library/bdBank/Deferred/User.php]: This query may get quite expensive. Have you tested with enough data?

File: library/bdBank/Installer.php [hunk 2]
```diff
@@ -67,14 +78,25 @@ class bdBank_Installer
             'field' => 'bdbank_money',
             'checkQuery' => 'SHOW COLUMNS FROM `xf_user` LIKE \'bdbank_money\'',
             'addQuery' => 'ALTER TABLE `xf_user` ADD COLUMN `bdbank_money` DECIMAL(13,4) DEFAULT \'0\'',
+            'modifyQuery' => 'ALTER TABLE `xf_user` MODIFY COLUMN `bdbank_money` DECIMAL(13,4) DEFAULT \'0\'',
             'dropQuery' => 'ALTER TABLE `xf_user` DROP COLUMN `bdbank_money`',
         ),
+        array(
+            'table' => 'xf_user',
+            'tableCheckQuery' => 'SHOW TABLES LIKE \'xf_user\'',
+            'field' => 'bdbank_credit',
```
Comment (daohoangson) [inline library/bdBank/Installer.php:76]: Field should be something adjustment I think? Credit seems wrong...

File: library/bdBank/Model/Bank.php [hunk 3]
```diff
@@ -74,11 +76,25 @@ public function canRefund(array $transaction, array $viewingUser = null)
         return false;
     }
 
-    public function getActionBonus($action, $extraData = array())
+    public function getActionBonus($action, $extraData = array(), $actionDate = null)
```
Comment (daohoangson) [inline library/bdBank/Model/Bank.php]: Let's make `$actionDate` the second parameter without default value -> force all bonus code to provide a valid date. Maybe throw exception if action date is empty / not a number too.

**Review by daohoangson — COMMENTED** [review]
User rank may get demoted with this code, please double check:

# Context

Rank 1: 100p
Rank 2: 1000p
Post bonus: 10p
User: p=994, a=0 (994 point, 0 adjustment -> 994+0 -> rank 1)

# Steps

1. User post something, got 10p -> p=1004, a=0 -> 1004+0 -> rank 2
2. Admin adjust post bonus from 10 -> 5, user has p=1004, a=-5, 1004-5=999 -> rank 1

**Comment by datbth** [comment]
> User rank may get demoted with this code, please double check:
> 
> # Context
> Rank 1: 100p
> Rank 2: 1000p
> Post bonus: 10p
> User: p=994, a=0 (994 point, 0 adjustment -> 994+0 -> rank 1)
> 
> # Steps
> 1. User post something, got 10p -> p=1004, a=0 -> 1004+0 -> rank 2
> 2. Admin adjust post bonus from 10 -> 5, user has p=1004, a=-5, 1004-5=999 -> rank 1

`credit = outgoingAdjustment - incomingAdjustment` anh, so if adjustment = -5, credit will be 5  
https://github.com/xfrocks/bdBank/pull/4/files#diff-7f32213b354a811b0aa84dfd57a8d964R42

File: library/bdBank/Model/Bank.php (hunk 3 above)
Comment (datbth) [reply library/bdBank/Model/Bank.php]: ok anh

File: library/bdBank/Installer.php (hunk 2 above)
Comment (datbth) [reply library/bdBank/Installer.php:76]: credit is the negated sum of adjustments anh.  
Also this makes more sense for addons like TinhteRewardStore to add credit for each transaction made. For example, exchanging 5 bdbank_money for a reward will grant 5 credit points

Comment (daohoangson) [reply library/bdBank/Installer.php:76]: > credit = outgoingAdjustment - incomingAdjustment anh, so if adjustment = -5, credit will be 5

Sounds too complicated eh. How about keeping track of adjustment the normal way (income - outgoing, adjustment will be -5) and use `money - adjustment` (money minus adjustment) instead of `money + credit` (money plus credit) for user group promotion?

**Comment by daohoangson** [comment]
Hmm, how about these:

## Context
Rank 1: 100p
Rank 2: 1000p
Post bonus: 10p
User: p=994, c=0 (994 point, 0 credit -> 994+0 -> rank 1)

## Steps

1. User post something, got 10p -> p=1004, c=0 -> 1004+0 -> rank 2
2. Admin adjust post bonus from 10 -> 5, user has p=999, c=5 -> 999+5=1004 -> rank 2
3. Post is deleted, user has p=994, c=5 -> 994+5=999-> rank 1
4. User post something else, got 5p -> p=999, c=5 -> 999+5=1004 -> rank 2

Basically user gets an extra +5 for nothing by deleting the old post?

File: library/bdBank/Installer.php (hunk 2 above)
Comment (datbth) [reply library/bdBank/Installer.php:76]: Because I think `credit` is more generic anh.  
It is more straight-forward and makes more sense for other addons like TinhteRewardStore to extend the calculation of `credit`, compared to `adjustment`.

**Comment by datbth** [comment]
Hmm looks like I need to do either of both of these:
* Remove related adjustments when a transaction is reversed
* During calculation of bdbank_money and bdbank_credit, only calculate the adjustments that belong to existing transactions.

File: library/bdBank/Installer.php (hunk 2 above)
Comment (daohoangson) [reply library/bdBank/Installer.php:76]: Other add-ons still need to insert record into adjustment table eh? So they have to understand the adjust concept anyway...

Oh, you plan to extend `_calculateAdditionalCredit` in those add-on instead of adding adjustment directly? That means there will be more tables involved in calculating each user point? It's currently quite complicated already, I'm not sure it's a good idea to stretch it even more.

**Comment by daohoangson** [comment]
> During calculation of bdbank_money and bdbank_credit, only calculate the adjustments that belong to existing transactions.

This may be hard to do, especially because transactions are only reverted (not deleted). 

> Remove related adjustments when a transaction is reversed

I'm not sure about this. Maybe it's better to add *another* adjustment to cancel out the existing adjustment? Deleting transaction data makes it hard to keep track of.

File: library/bdBank/Installer.php (hunk 2 above)
Comment (datbth) [reply library/bdBank/Installer.php:76]: yes anh, I plan to extend `_calculateAdditionalCredit`  in other addons. This only affects users' promotion so I think it's fine?  
I think it's kinda wrong for TinhteRewardStore to create an adjustment record for each of its transaction

File: library/bdBank/Model/Bank.php (hunk 3 above)
Comment (datbth) [reply library/bdBank/Model/Bank.php]: ah but this signature changing can affect other addons anh, because current second parameter `$extraData` is optional

Comment (daohoangson) [reply library/bdBank/Model/Bank.php]: Yeah, it's the point. This is a major / breaking change so all dependencies should be updated to work correctly or stop working.

File: library/bdBank/Installer.php (hunk 2 above)
Comment (daohoangson) [reply library/bdBank/Installer.php:76]: What is the implementation plan for the other add-on? 

## A. User purchase flow

1. User decides to buy something
1. Add-on insert transaction record using `bdBank_Model_Personal::transfer` (e.g. transaction from user -> bank, amount=100p) -> bdBank update `bdbank_money` field
1. Add-on insert purchase record in its own table and update `bdbank_credit` by itself  (e.g. bdbank_credit = bdbank_credit + 100)

## B. User rebuild flow

1. Add-on override `_calculateAdditionalCredit` to return total purchase amount from its own table

A3 sounds pretty crazy to me.

Comment (datbth) [reply library/bdBank/Installer.php:76]: Yes the above is what I'm planning to do.  
Could you clarify your concerns anh? :<  

In A3, the logic to update `bdbank_credit` might be implemented by bdBank itself.
Does it make sense to:
* make a new transaction type, say `bdBank_Model_Bank::TYPE_CREDITABLE`
* when a transaction of this new type is made, the transaction's amount is negated and added to from_user's credit

File: library/bdBank/Deferred/User.php (hunk 1 above)
Comment (datbth) [reply library/bdBank/Deferred/User.php]: hmm all `comment`, `from_user_id` and `to_user_id` columns are indexed (except `archive.from_user_id`, don't know why?).  
Also this operation is rarely executed so I think it's fine?  

Besides, not sure if the following is better:
```sql
SELECT SUM(adj.amount)
FROM xf_bdbank_transaction_adjustment adj
WHERE adj.comment IN (
    SELECT t.comment
    FROM xf_bdbank_transaction t
    WHERE t.$userIdCol = $userId
      AND t.reversed = 0
    UNION ALL
    SELECT a.comment
    FROM xf_bdbank_archive a
    WHERE a.$userIdCol = $userId
)
```

File: library/bdBank/Installer.php (hunk 2 above)
Comment (datbth) [reply library/bdBank/Installer.php:76]: If we go with the `TYPE_CREDITABLE`, the other addons would also **not** need to override `_calculateAdditionalCredit`

**Comment by datbth** [comment]
> > During calculation of bdbank_money and bdbank_credit, only calculate the adjustments that belong to existing transactions.
> 
> This may be hard to do, especially because transactions are only reverted (not deleted).

Just realized I've already been doing this, which raises the above concern about performance.

> > Remove related adjustments when a transaction is reversed
> 
> I'm not sure about this. Maybe it's better to add _another_ adjustment to cancel out the existing adjustment? Deleting transaction data makes it hard to keep track of.

Added logic that makes new adjustments to cancel out the adjustments of reversed/deleted transactions via https://github.com/xfrocks/bdBank/pull/4/commits/b9cd50b43fa64b124cdcb9ba470cc224095a866a

File: library/bdBank/Model/Bank.php (hunk 3 above)
Comment (datbth) [reply library/bdBank/Model/Bank.php]: hmm looks like this won't work for attachments. Currently, we do not make transactions for individual attachments (`bdBank_Model_Bank#macro_bonusAttachment`).  
While it also does not make sense to use the creation date of the content (that the attachments belong to) as the actionDate, does it?

File: library/bdBank/Deferred/User.php (hunk 1 above)
Comment (daohoangson) [reply library/bdBank/Deferred/User.php]: I think the problematic table is adjustment. It doesn't have any index so when you left join it will scan all adjustments. We will have millions of those -> it will be slow.

The sub query in WHERE looks better I think. I do recall some issues with sub query though, maybe we need to benchmark this with real data. How many rows do you have?

File: library/bdBank/Installer.php (hunk 2 above)
Comment (daohoangson) [reply library/bdBank/Installer.php:76]: If you let bdBank update `bdbank_credit` then the other add-on doesn't need to care about credit / adjustment -> we shouldn't introduce 2 completely new concepts in this PR (adjustment and credit). My point is: stick to one, either use adjustment for everything or just use credit everywhere. Currently it's quite confusing, adjustment in some and credit in others...

Regarding the new transaction type, I think that's great. Let's do that.

> when a transaction of this new type is made, the transaction's amount is negated and added to from_user's credit

I think we should do 3 things:

- Insert transaction record
- Insert adjustment record
- Update xf_user fields

All in the same db transaction.

File: library/bdBank/Model/Bank.php (hunk 3 above)
Comment (daohoangson) [reply library/bdBank/Model/Bank.php]: Let's drop attachment bonuses completely then.

File: library/bdBank/Deferred/User.php (hunk 1 above)
Comment (datbth) [reply library/bdBank/Deferred/User.php]: I added an index on adjustment.comment anh.  
Let me do some benchmarkings

File: library/bdBank/Model/Bank.php (hunk 3 above)
Comment (datbth) [reply library/bdBank/Model/Bank.php]: done via https://github.com/xfrocks/bdBank/pull/4/commits/d8c34a8b4825220f0312b99e205f347515520321

File: library/bdBank/Installer.php (hunk 2 above)
Comment (datbth) [reply library/bdBank/Installer.php:76]: hmm I don't want to stretch the meaning of adjustments. It is weird to insert a transaction and insert an adjustment for that new transaction at the same time.  

I think the meaning of adjustment and credit can be summarized like this:
* An adjustment is a modification to a transaction's amount and is made by the system. It affects users' bdbank_money
* Credits are the points that are only considered when evaluating a user's rank. Those points are granted by adjustments or creditable transactions. Credits do not affect bdbank_money

Comment (datbth) [reply library/bdBank/Installer.php:76]: Using credit for everything might make more sense than using adjustment for everything, in terms of meaning.  
Let me try this approach.

Comment (datbth) [reply library/bdBank/Installer.php:76]: Ah no,
* using credit for everything: cannot distinguish between credits made adjustments and credits made by creditable transactions. The former is counted towards bdbank_money why the latter is not.
* using adjustment for everything: cannot distinguish between adjustments made by actual system adjustments and adjustments made by creditable transactions. The former is counted towards bdbank_money why the latter is not.

-> Need to distinguish credit and adjustment for bdbank_money calculation

Comment (daohoangson) [reply library/bdBank/Installer.php:76]: Yeah, if merge two concepts together, a slight change is required during rebuilding:

- Currently (in this PR): rebuilder will insert new adjustment record
- Required change: rebuilder should add insert 1 transaction record and 1 adjustment record

Example: post bonus from 10 -> 5.

- Currently (in this PR): rebuilder will insert adjustment record of -5 (target 5 - current 10 = -5)
- Required change: rebuilder will insert transaction -5 **and** adjustment +5.

Benefits:

- Only one new concept
- Simplify `bdbank_money` calculation: just sum(amount) of all transaction
- Easy to calculate `bdbank_credit` (or `bdbank_adjustment`, up to you): sum(amount) of all adjustment

Thinking about this, is it possible to use transaction table for adjustment too? The adjustment amount is always negative of the transaction amount, if we set `transaction_type=adjustment` and take it into account while calculating `bdbank_money` and `bdbank_credit`, that will be fast.

Comment (datbth) [reply library/bdBank/Installer.php:76]: implemented the above approach via https://github.com/xfrocks/bdBank/pull/4/commits/360985f6504bb8c1aba0551adbc73e119cffaaa5: 
* track adjustments using `transaction` table
* create `credit` record for every adjustment/creditable transaction, by using a DB trigger. Without the trigger, we cannot know which credit is made for which transaction. One drawback is that it will be difficult  to update the trigger definition, we will probably need to lock bdbank_transaction table to do so.
* update `xf_user.bdbank_credit` every time a `credit` record is inserted, by using a DB trigger. Trigger is necessary because of the above trigger. Besides, the trigger reduces code complexity and ensures we do not miss updating bdbank_credit
* rebuild `xf_user.bdbank_money` like before
* rebuild `xf_user.bdbank_credit` = SUM(amount) of credit records of that user

Comment (daohoangson) [reply library/bdBank/Installer.php:76]: Remind me why you still need the `xf_bdbank_credit` table?

Comment (datbth) [reply library/bdBank/Installer.php:76]: The triggers could be removed if `lastInsertId` is thread-safe, but I'm not sure if it is?

Comment (daohoangson) [reply library/bdBank/Installer.php:76]: It is because everything is in a transaction.

However, I don't think you need that table at all... Why all the hassle?

Comment (datbth) [reply library/bdBank/Installer.php:76]: ah right...

Comment (datbth) [reply library/bdBank/Installer.php:76]: updated to use transactions to track everything: https://github.com/xfrocks/bdBank/compare/360985f..dd1fa99

Comment (datbth) [reply library/bdBank/Installer.php:76]: Hmm just found 1 issue:
1. User makes 1 post and is awarded 1 point (transaction 1 with amount 1) -> bdbank_money = 1
2. Admin adjusts the bonus for posts to 3 points
3. Admin rebuilds bonuses, which creates new ADJUSTMENT transaction 2 with amount 2. Note that at this point bdbank_money is still 1
4. User deletes the post -> system tries to revert transactions 1 and 2, but will only be able to revert one of them, because user doesn't have sufficient money

Possible solutions:
1. Stop trying to make a dummy transaction to validate the reverse (https://github.com/xfrocks/bdBank/pull/4/files#diff-f555f4f5a103abc1dca15df0eb138b7dR381)
2. Make sure bdbank_money is updated within the same transaction with bonus rebuilding

Comment (daohoangson) [reply library/bdBank/Installer.php:76]: After step (3), user values will be money=1, credit=2 -- is this correct?

If it is, when step (4) happens, reverting t1 will change money=0, credit=2 then reverting t2 will change money=0, credit=0 -- I think?

> Make sure bdbank_money is updated within the same transaction with bonus rebuilding

This cannot be done because rebuilding is already very slow. If we update `xf_user` during this time, it will be even slower...

Comment (datbth) [reply library/bdBank/Installer.php:76]: > After step (3), user values will be money=1, credit=2 -- is this correct?  

Yes anh  

> If it is, when step (4) happens, reverting t1 will change money=0, credit=2 then reverting t2 will change money=0, credit=0 -- I think?

Reverting t2 would be ignored because making the dummy transaction would fail. The solution 1 above tries to solve this

Comment (daohoangson) [reply library/bdBank/Installer.php:76]: > Reverting t2 would be ignored because making the dummy transaction would fail.

I think it's a bug in the test-transaction code because it didn't take into account transaction type. e.g. only verify if type=system or type=user?

Comment (datbth) [reply library/bdBank/Installer.php:76]: Not sure if it's safe to verify for certain types only? What is the reason for making a test transaction when reverting anh, and it is a good idea to ignore the fail test transactions?

Comment (daohoangson) [reply library/bdBank/Installer.php:76]: I have taken a closer look, it is ignoring transaction exception already eh?

Comment (datbth) [reply library/bdBank/Installer.php:76]: Yeah it is ignoring anh, but I'm questioning whether ignoring is good

Comment (daohoangson) [reply library/bdBank/Installer.php:76]: I don't recall the reasoning but if it was ignoring already then let keep it this way, at least for now.

Comment (datbth) [reply library/bdBank/Installer.php:76]: Atm the exception is ignored but the the transaction won't be reversed.  
So should I make the transaction testing ignore TYPE_ADJUSTMENT transactions?

Comment (daohoangson) [reply library/bdBank/Installer.php:76]: Ah... I see.

No, don't ignore by type for now. Adjustment will affect `bdbank_money` and should be tested by the same logic.

> Atm the exception is ignored but the the transaction won't be reversed.

Sounds dangerous to change this logic, it may have some serious consequence that we don't know for now. Let's create a new issue to look into it later and leave it out of this PR.

Comment (datbth) [reply library/bdBank/Installer.php:76]: Ok anh, I will create the issue.  
Please help review this PR again nha

File: library/bdBank/Deferred/User.php [hunk 4]
```diff
@@ -69,4 +75,26 @@ protected function _fetchSum($formula, $where)
 
         return bdBank_Helper_Number::add($liveValue, $archiveValue);
     }
+
+    protected function _fetchCredit($formula, $where)
```
Comment (daohoangson) [inline library/bdBank/Deferred/User.php]: Let's modify `_fetchSum` and remove `_fetchCredit`. I think we can do something like:

```sql
SELECT transaction_type, SUM(amount), SUM(tax_amount)
FROM xf_bdbank_transaction
WHERE to_user_id = 123 AND reversed= 0
GROUP BY transaction_type
```

Then do some more calculation in PHP according to transaction type.

File: library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php [hunk 5]
```diff
@@ -18,14 +18,12 @@ public function bdBank_doBonus()
...
-            $point = $bank->getActionBonus($bonusType, array('forum' => $forum));
+            $point = $bank->getActionBonus($bonusType, $this->get('post_date'), array('forum' => $forum));
         }
 
         if (!$this->isInsert()) {
-            $reverseResult = $bank->reverseSystemTransactionByComment($comment, $point);
-            if (count($reverseResult['skipped']) > 0) {
-                return;
-            }
+            $bank->makeTransactionAdjustments($comment, $point);
```
Comment (daohoangson) [inline library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: Let's use the old reverse+rebonus logic here. Because if a post is deleted, the old credit should be forfeited.

**Review by daohoangson — CHANGES_REQUESTED** [review]

File: library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php (hunk 5 above)
Comment (datbth) [reply library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: done via 5feab2c95482555382fcfe35849cea64274bfb9a

File: library/bdBank/Deferred/User.php (hunk 4 above)
Comment (datbth) [reply library/bdBank/Deferred/User.php]: done via d1e7cd33149f80d29e4eae373a67e8a156360a0d

File: library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php (hunk 5 above)
Comment (datbth) [reply library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: ah but then bonus adjustment for post won't grant credit.  
Need to use separate logic to rebuild bonus for posts instead of reusing `bdBank_doBonus` then?

Comment (datbth) [reply library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: Or check if post is actually deleted in `bdBank_doBonus`

Comment (datbth) [reply library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: fixed in 84ac3d354bf31dfc0954c9655592a5ddac5a803a

Comment (daohoangson) [reply library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: > ah but then bonus adjustment for post won't grant credit.

That's the correct behavior I think. If user deleted the post -> all previous bonus/credit transactions are reverted. When it's undeleted -> system send out new bonus, no credit/adjustment logic is required. 

I think the changes in 84ac3d3 should be reverted.

Comment (datbth) [reply library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: I'm referring to when we adjust the bonus points for post, anh.  
Example:
1 .adjust bonus for  post from 2 to 1
2. rebuild bonuses
3. if we reverse old transaction and create new one with 1 point, user won't get 1 point of credit

Comment (datbth) [reply library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: if the post is soft-deleted, `checkPostQuality` would return false.  
if the post is hard-deleted, rebuild bonuses won't process the post.

Update: ah so we during rebuilding, we still need to check if post is soft-deleted

Comment (datbth) [reply library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: ad17c95

Comment (daohoangson) [reply library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: Hmm, maybe for rebuilding, we don't need to check `XenForo_Model_Post:: isDeleted`. Just adjust it to `$point`, always?

Comment (datbth) [reply library/bdBank/XenForo/DataWriter/DiscussionMessage/Post.php:25]: ah right, reverting ad17c95 now

**Review by daohoangson — APPROVED** [review]
